shared/modules/object.utils.ts

```diff
@@ -0,0 +1,326 @@
+    // Mask the entire array
+    return 'array';
+  }
+
+  // Consult the mask to know which properties to mask within the object
+  const effectiveMask = isMaskWithAllProperties(mask)
+    ? mask[AllProperties]
+    : mask;
+  return array.map((item) => maskValue(item, effectiveMask));
+}
+
+/**
+ * Ensures that a mask for a plain object is valid.
+ * @param mask The mask to validate.
+ * @param useStrictMaskValidation - If true, only allow mask to be a plain
+ * object. If false, allow it to be a plain object but also boolean or
+ * undefined.
+ * @throws If the mask is not valid.
+ */
+function validatePlainObjectMask(
+  mask: Mask,
+  useStrictMaskValidation: boolean,
+): asserts mask is PlainObjectMask {
+  if (useStrictMaskValidation) {
+    if (!isPlainObject(mask)) {
+      throw new Error('The mask for a plain object must be a plain object');
+    }
+  } else if (
+    !(typeof mask === 'boolean' || isPlainObject(mask) || mask === undefined)
+  ) {
+    throw new Error(
+      'The mask for a plain object must be a boolean, a plain object, or undefined',
+    );
+  }
+
+  if (isMaskWithAllProperties(mask)) {
+    validateMaskWithAllProperties(mask);
+  }
+}
+
+/**
+ * Masks a plain object.
+ *
+ * @param plainObject - The object to mask.
+ * @param mask - The mask to apply to the object.
+ * @returns The masked object.
+ */
+function maskPlainObject(
+  plainObject: MaskablePlainObject,
+  mask: PlainObjectMask,
+): MaskedPlainObject {
+  if (mask === true) {
+    // Don't mask any properties in the object
+    return plainObject;
+  }
+
+  if (mask === undefined || mask === false) {
+    // Mask the entire object
+    return 'object';
+  }
+
+  // Consult the mask to know which properties to mask within the object
+  return Object.entries(plainObject).reduce<
+    Exclude<MaskedPlainObject, 'object'>
+  >((obj, [key, value]) => {
+    const maskForKey = isMaskWithAllProperties(mask)
+      ? mask[AllProperties]
+      : mask[key];
+    if (maskForKey === true) {
+      // Don't mask the value
+      return { ...obj, [key]: value };
+    } else {
+      // Mask the value
+      return { ...obj, [key]: maskValue(value, maskForKey) };
+    }
+  }, {});
+}
+
+/**
+ * Mask anything that is JSON-serializable. Called recursively to mask
+ * sub-values.
+ *
+ * @param value - The value to mask.
+ * @param mask - The mask to apply to the value.
+ * @returns The masked value.
+ * @throws If mask is not a boolean, a plain object, or undefined.
+ */
+function maskValue(value: MaskableValue, mask: Mask): MaskedValue {
+  if (
+    typeof value === 'string' ||
+    typeof value === 'number' ||
+    typeof value === 'boolean' ||
+    value === null ||
+    value === undefined
+  ) {
+    if (!(typeof mask === 'boolean' || mask === undefined)) {
+      throw new Error(
+        'The mask for a primitive must be a boolean or undefined',
+      );
+    }
+
+    return maskPrimitive(value, mask);
+  }
+
+  if (Array.isArray(value)) {
```

Comment:
This is where support was added for arrays. `maskValue` is called recursively as the source object is iterated over and descended into, so eventually this will be called.